core/src/main/java/com/linecorp/armeria/server/VirtualHost.java

```diff
@@ -348,15 +354,46 @@ public boolean shutdownAccessLogWriterOnStop() {
     }
 
     /**
-     * Finds the {@link Service} whose {@link Router} matches the {@link RoutingContext}.
+     * Finds the {@link HttpService} whose {@link Router} matches the {@link RoutingContext}.
      *
-     * @param routingCtx a context to find the {@link Service}.
+     * @param routingCtx a context to find the {@link HttpService}.
      *
      * @return the {@link ServiceConfig} wrapped by a {@link Routed} if there's a match.
      *         {@link Routed#empty()} if there's no match.
      */
     public Routed<ServiceConfig> findServiceConfig(RoutingContext routingCtx) {
-        return router.find(requireNonNull(routingCtx, "routingCtx"));
+        return findServiceConfig(routingCtx, false);
+    }
+
+    /**
+     * Finds the {@link HttpService} whose {@link Router} matches the {@link RoutingContext}.
+     *
+     * @param routingCtx a context to find the {@link HttpService}.
+     * @param useFallbackService whether to use the fallback {@link HttpService} when there is no match.
+     *                           If {@code true}, the returned {@link Routed} will always be present.
+     *
+     * @return the {@link ServiceConfig} wrapped by a {@link Routed} if there's a match.
+     *         {@link Routed#empty()} if there's no match.
+     */
+    public Routed<ServiceConfig> findServiceConfig(RoutingContext routingCtx, boolean useFallbackService) {
+        final Routed<ServiceConfig> routed = router.find(requireNonNull(routingCtx, "routingCtx"));
+        if (routed.isPresent() || !useFallbackService) {
+            return routed;
+        }
+
+        if (routingCtx.isCorsPreflight()) {
```

Comment:
Is there a reason to have this logic here and not in the fallback service?

Reply:
Brainfart 2

Reply:
Moved